File: CRM-Server/app/schemas/approval_generic.py
"""通用审批 API 端点请求/响应 Schema（Task A6）。

与 `app/schemas/approval.py` 解耦：旧 schema 仍服务 `/contracts/{contract_id}/...`
合同专用端点（含 ApprovalDetailResponse 等完整响应模型），本模块仅定义通用端点
`/v1/approvals/{entity_type}/{entity_id}/...` 所需的轻量请求体。
`ApprovalActionRequest`（action/comment/updated_time）复用既有
`schemas.approval.ApprovalActionRequest`，避免重复定义与 enum 行为分叉。

设计要点：
- submit 仅需 comment（可选）
- approve 复用 schemas.approval.ApprovalActionRequest（ApprovalActionEnum + 乐观锁 updated_time）
"""
from typing import Optional, List

from pydantic import BaseModel, Field


class ApprovalSubmitRequest(BaseModel):
    """通用审批提交请求。"""
    comment: Optional[str] = Field(
        None,
        description="提交说明，提交时填写的说明信息（可选）",
    )


class GenericApprovalSubmitResponse(BaseModel):
    """通用审批提交响应。"""
    approval_id: int = Field(..., description="新建审批实例 ID")
    status: str = Field(..., description="审批状态：PENDING/APPROVED/REJECTED/CANCELLED")


class ApprovalCustomerInfoResponse(BaseModel):
    """审批关联客户/公司基础信息。"""

    id: int = Field(..., description="客户ID")
    account_name: str = Field(..., description="客户公司名称")
    industry: Optional[str] = Field(None, description="所属行业")
    city: Optional[str] = Field(None, description="所在城市")
    company_scale: Optional[str] = Field(None, description="公司规模")
    source: Optional[str] = Field(None, description="客户来源")
    status: Optional[int] = Field(None, description="客户状态")
    owner_id: Optional[str] = Field(None, description="客户负责人ID")


# ============================================================================
# Task C3：通用审批列表端点 GET /v1/approvals 响应模型
# ============================================================================
# 对齐前端 `schemas/approvalGeneric.ts` 的 ApprovalListItemSchema / ApprovalListResponseSchema：
#   - items: 审批实例行（含按 business_type 内存 join 出的 application_number /
#     entity_name / entity_amount 三摘要字段，由 CRUD 层批量预取避免 N+1）
#   - overdue_hours: 后端 SQL/Python 计算（now - created_time，单位小时），
#     仅 PENDING 行有意义；SUBMITTED/PROCESSED tab 也回传，前端按需展示
#   - pending_count: 不论请求哪个 tab，后端都附当前用户「待我审批」总数，供侧边栏徽章
# ============================================================================


class ApprovalListItemResponse(BaseModel):
    """通用审批列表项（GET /v1/approvals items 元素）。"""

    id: int = Field(..., description="审批实例ID")
    business_type: str = Field(..., description="业务单据类型：CONTRACT / PAYMENT / INVOICE / LICENSE / OPPORTUNITY")
    business_id: int = Field(..., description="业务单据ID")
    application_number: str = Field(..., description="单号（合同 contract_number / 发票 application_number / 回款合成 PAY-{id}）")
    entity_name: Optional[str] = Field(None, description="客户/实体摘要（合同 contract_name / 发票 invoice_title_text；回款暂为 None）")
    entity_amount: Optional[float] = Field(None, description="单据金额（合同 total_amount / 发票 invoice_amount / 回款 actual_amount）")
    actual_payer_name: Optional[str] = Field(None, description="实际付款方名称，仅回款审批返回")
    license_status: Optional[str] = Field(None, description="License 申请状态，仅 License 审批返回")
    customer_info: Optional[ApprovalCustomerInfoResponse] = Field(None, description="关联客户/公司基础信息")
    submitter_id: str = Field(..., description="提交人系统用户ID")
    submitter_name: Optional[str] = Field(None, description="提交人姓名")
    status: str = Field(..., description="审批状态：PENDING / APPROVED / REJECTED / CANCELLED")
    created_time: str = Field(..., description="提交时间 ISO8601")
    updated_time: str = Field(..., description="最后更新时间 ISO8601")
    overdue_hours: Optional[int] = Field(None, description="超时小时数（now - created_time），仅 PENDING 行有意义")


class ApprovalGenericListResponse(BaseModel):
    """通用审批列表响应（GET /v1/approvals）。"""

    items: List[ApprovalListItemResponse] = Field(..., description="审批列表项")
    total: int = Field(..., ge=0, description="当前 tab 过滤后的总数")
    pending_count: int = Field(..., ge=0, description="当前用户的「待我审批」总数，任意 tab 响应都携带，供侧边栏徽章")

File: CRM-Client/src/schemas/approvalGeneric.ts
/**
 * Zod Schema - 通用审批（Generic Approval）类型
 *
 * @description 映射后端通用审批端点 `app/api/approvals.py` 的
 * `_serialize_generic_approval` 输出与 `app/schemas/approval_generic.py` 的
 * 请求/响应模型。覆盖 CONTRACT / PAYMENT / INVOICE / LICENSE 四类业务单据。
 *
 * 字段命名沿后端 snake_case 约定（与既有 schemas/lead.ts、common.ts 一致）。
 * datetime 字段使用 `z.string().min(1)` 而非 `.datetime()`：后端 SQLAlchemy
 * `DateTime`（无 tzinfo）经 FastAPI 序列化为 `2026-07-01T10:00:00`（无 Z 后缀），
 * `z.string().datetime()` 强制要求 RFC 3339 时区段会拒绝真实响应。
 */

import { z } from 'zod'

// ===== 业务单据类型（A1 BusinessType） =====
export const EntityTypeSchema = z.enum(['CONTRACT', 'PAYMENT', 'INVOICE', 'LICENSE'])
export type EntityType = z.infer<typeof EntityTypeSchema>

// ===== 审批动作枚举（ApprovalActionEnum） =====
export const ApprovalActionSchema = z.enum(['APPROVE', 'REJECT', 'ROLLBACK'])
export type ApprovalAction = z.infer<typeof ApprovalActionSchema>

// ===== 审批状态枚举（ApprovalStatusEnum） =====
export const ApprovalStatusEnum = z.enum(['PENDING', 'APPROVED', 'REJECTED', 'CANCELLED'])
export type ApprovalStatus = z.infer<typeof ApprovalStatusEnum>

export const LicenseApplicationStatusSchema = z.enum(['DRAFT', 'PENDING_REVIEW', 'APPROVED', 'REJECTED', 'ISSUED'])
export type LicenseApplicationStatus = z.infer<typeof LicenseApplicationStatusSchema>

// ===== 审批人状态（UserStatus，可空） =====
export const ApproverStatusSchema = z.enum(['active', 'inactive', 'suspended'])
export type ApproverStatus = z.infer<typeof ApproverStatusSchema>

// ===== 关联客户/公司基础信息 =====
export const ApprovalCustomerInfoSchema = z.object({
  id: z.number().int().positive(),
  account_name: z.string().min(1),
  industry: z.string().nullable().optional(),
  city: z.string().nullable().optional(),
  company_scale: z.string().nullable().optional(),
  source: z.string().nullable().optional(),
  status: z.number().int().nullable().optional(),
  owner_id: z.string().nullable().optional()
})
export type ApprovalCustomerInfo = z.infer<typeof ApprovalCustomerInfoSchema>

// ===== 审批记录（timeline 节点） =====
export const ApprovalRecordSchema = z.object({
  id: z.number().int().positive(),
  approval_id: z.number().int().positive(),
  node_id: z.number().int().nullable(),
  node_name: z.string().nullable(),
  approver_id: z.string().nullable(),
  approver_name: z.string().nullable(),
  approver_status: ApproverStatusSchema.nullable(),
  approver_status_display: z.string().nullable(),
  action: z.enum(['SUBMIT', 'APPROVE', 'REJECT', 'ROLLBACK']).nullable(),
  comment: z.string().nullable(),
  created_time: z.string().min(1)
})

export type ApprovalRecord = z.infer<typeof ApprovalRecordSchema>

// ===== 审批详情响应 =====
export const ApprovalDetailSchema = z.object({
  id: z.number().int().positive(),
  business_type: EntityTypeSchema,
  business_id: z.number().int().positive(),
  contract_id: z.number().int().nullable(),
  flow_id: z.number().int(),
  flow_name: z.string().nullable(),
  current_node_id: z.number().int().nullable(),
  current_node_name: z.string().nullable(),
  status: ApprovalStatusEnum,
  submitter_id: z.string().min(1),
  submitter_name: z.string().nullable(),  // 允许 null（发票/回款申请人无姓名字段）
  created_time: z.string().min(1),
  updated_time: z.string().min(1),
  flow_is_active: z.boolean().nullable(),
  flow_disabled_warning: z.string().nullable(),
  customer_info: ApprovalCustomerInfoSchema.nullable().optional(),
  records: z.array(ApprovalRecordSchema),
  // Task 6: 发票文件上传字段（仅 INVOICE 类型）
  invoice_file_path: z.string().nullable().optional(),
  invoice_number: z.string().nullable().optional(),
  issued_time: z.string().nullable().optional(),
  // License 发放状态字段（仅 LICENSE 类型）
  license_status: LicenseApplicationStatusSchema.nullable().optional()
})

export type ApprovalDetail = z.infer<typeof ApprovalDetailSchema>

// ===== 通用提交响应（GenericApprovalSubmitResponse） =====
export const ApprovalSubmitResponseSchema = z.object({
  approval_id: z.number().int().nonnegative(),
  status: ApprovalStatusEnum
})

export type ApprovalSubmitResponse = z.infer<typeof ApprovalSubmitResponseSchema>

// ===== 撤回响应（MessageResponse） =====
export const MessageResponseSchema = z.object({
  message: z.string().min(1)
})

export type MessageResponse = z.infer<typeof MessageResponseSchema>

// ===== 批量审批失败条目（BulkApproveFailedItem） =====
export const BulkApproveFailedItemSchema = z.object({
  id: z.number().int(),
  reason: z.string().min(1)
})

export type BulkApproveFailedItem = z.infer<typeof BulkApproveFailedItemSchema>

// ===== 批量审批汇总响应（BulkApproveResponse） =====
export const BulkApproveResponseSchema = z.object({
  success_count: z.number().int().nonnegative(),
  failed: z.array(BulkApproveFailedItemSchema)
})

export type BulkApproveResponse = z.infer<typeof BulkApproveResponseSchema>

// ===== FinanceApprovalCenter 列表项（Task C3 / E2 角色过滤） =====
// 后端列表端点按 tab+business_type 严格按角色过滤（E2）：
//   submitted  → submitter_id == current_user AND team_id
//   pending    → current_node.approve_role IN (user_roles) AND status=PENDING AND team_id
//   processed  → records.approver_id == current_user AND team_id
// 单号 / 实体摘要 / 金额由后端按 business_type 分组批量预取后内存 join（E9）。
export const ApprovalTabSchema = z.enum(['pending', 'processed', 'submitted'])
export type ApprovalTab = z.infer<typeof ApprovalTabSchema>

export const ApprovalListItemSchema = z.object({
  id: z.number().int().positive(),
  business_type: EntityTypeSchema,
  business_id: z.number().int().positive(),
  application_number: z.string().min(1),
  entity_name: z.string().nullable(),
  entity_amount: z.number().nullable(),
  actual_payer_name: z.string().nullable().optional(),
  license_status: LicenseApplicationStatusSchema.nullable().optional(),
  customer_info: ApprovalCustomerInfoSchema.nullable().optional(),
  submitter_id: z.string().min(1),
  submitter_name: z.string().nullable(),  // 允许 null（发票/回款申请人无姓名字段）
  status: ApprovalStatusEnum,
  created_time: z.string().min(1),
  updated_time: z.string().min(1),
  overdue_hours: z.number().nullable()
})
export type ApprovalListItem = z.infer<typeof ApprovalListItemSchema>

export const ApprovalListResponseSchema = z.object({
  items: z.array(ApprovalListItemSchema),
  total: z.number().int().nonnegative(),
  pending_count: z.number().int().nonnegative()
})
export type ApprovalListResponse = z.infer<typeof ApprovalListResponseSchema>

export const ApprovalListQuerySchema = z.object({
  tab: ApprovalTabSchema,
  business_type: EntityTypeSchema.optional(),
  page: z.number().int().positive().default(1),
  page_size: z.number().int().positive().default(20)
})
export type ApprovalListQuery = z.infer<typeof ApprovalListQuerySchema>

// ===== 发票文件上传审批响应（Task 5） =====
export const ApproveWithFileResponseSchema = z.object({
  success: z.boolean(),
  message: z.string().min(1),
  file_path: z.string().min(1),
  invoice_number: z.string().nullable(),
  new_status: z.string().min(1)
})
export type ApproveWithFileResponse = z.infer<typeof ApproveWithFileResponseSchema>
